File: interfaces/webui/src/lib/api/folders.ts
/**
 * ETHAN WebUI — Folders API service
 *
 * Client passif des dossiers génériques Core (/v1/folders).  Toute la
 * logique (arborescence, relations multi-ressources, validation) appartient
 * au Core (core/folders) : cette couche n'affiche et ne transmet que des
 * intentions utilisateur.
 */

import { apiFetch } from '@/lib/api/client';

export type FolderResourceType = "knowledge" | "collection" | "skill";

export interface Folder {
	id: string;
	name: string;
	description: string;
	user_id: string;
	/** Dossier parent (arborescence libre créée par l'utilisateur). */
	parent_id: string | null;
	/**
	 * Collection Knowledge associée (contexte de navigation) — référence
	 * validée par le Core ; un dossier n'est PAS un répertoire de la collection.
	 */
	collection_id: string | null;
	icon: string | null;
	order: number;
	metadata: Record<string, unknown>;
	created_at: string;
	updated_at: string;
}

export interface FolderTree extends Folder {
	/** Nombre de ressources directement classées dans le dossier. */
	resource_count: number;
	children: FolderTree[];
}

export interface FolderResource {
	resource_type: FolderResourceType;
	resource_id: string;
	/** Record réel résolu par le manager Core propriétaire. */
	record: Record<string, unknown> | null;
}

export async function listFolderTree(
	userId?: string,
	collectionId?: string | null,
): Promise<FolderTree[]> {
	const params = new URLSearchParams();
	if (userId) params.set("user_id", userId);
	if (collectionId) params.set("collection_id", collectionId);
	const query = params.toString() ? `?${params.toString()}` : "";
	return apiFetch<FolderTree[]>(`/v1/folders/tree${query}`);
}

export async function getFolder(folderId: string): Promise<Folder> {
	return apiFetch<Folder>(`/v1/folders/${folderId}`);
}

export async function createFolder(data: {
	name: string;
	description?: string;
	user_id?: string;
	parent_id?: string | null;
	collection_id?: string | null;
	icon?: string | null;
	order?: number;
}): Promise<Folder> {
	return apiFetch<Folder>("/v1/folders", {
		method: "POST",
		body: JSON.stringify(data),
	});
}

export async function updateFolder(
	folderId: string,
	data: {
		name?: string;
		description?: string;
		icon?: string | null;
		order?: number;
		parent_id?: string | null;
		collection_id?: string | null;
	},
): Promise<Folder> {
	return apiFetch<Folder>(`/v1/folders/${folderId}`, {
		method: "PATCH",
		body: JSON.stringify(data),
	});
}

export async function moveFolder(
	folderId: string,
	parentId: string | null,
): Promise<Folder> {
	return apiFetch<Folder>(`/v1/folders/${folderId}/move`, {
		method: "POST",
		body: JSON.stringify({ parent_id: parentId }),
	});
}

export async function deleteFolder(folderId: string): Promise<void> {
	await apiFetch(`/v1/folders/${folderId}`, { method: "DELETE" });
}

export async function listFolderResources(
	folderId: string,
	resourceType?: FolderResourceType,
): Promise<FolderResource[]> {
	const query = resourceType ? `?resource_type=${resourceType}` : "";
	return apiFetch<FolderResource[]>(`/v1/folders/${folderId}/resources${query}`);
}

export async function attachFolderResource(
	folderId: string,
	resourceType: FolderResourceType,
	resourceId: string,
): Promise<void> {
	await apiFetch(`/v1/folders/${folderId}/resources`, {
		method: "POST",
		body: JSON.stringify({ resource_type: resourceType, resource_id: resourceId }),
	});
}

export async function detachFolderResource(
	folderId: string,
	resourceType: FolderResourceType,
	resourceId: string,
): Promise<void> {
	await apiFetch(
		`/v1/folders/${folderId}/resources/${resourceType}/${resourceId}`,
		{ method: "DELETE" },
	);
}

/** Déplace une ressource vers l'ensemble de dossiers donné (remplacement ; [] = sans dossier). */
export async function moveResource(
	resourceType: FolderResourceType,
	resourceId: string,
	folderIds: string[],
): Promise<{ folder_ids: string[] }> {
	return apiFetch<{ folder_ids: string[] }>("/v1/folders/move-resource", {
		method: "POST",
		body: JSON.stringify({
			resource_type: resourceType,
			resource_id: resourceId,
			folder_ids: folderIds,
		}),
	});
}

/** Ressources d'un type classées dans aucun dossier. */
export async function listUntaggedResources(
	resourceType: FolderResourceType,
): Promise<Record<string, unknown>[]> {
	return apiFetch<Record<string, unknown>[]>(
		`/v1/folders/untagged?resource_type=${resourceType}`,
	);
}

/** Dossiers contenant une ressource (multi-membership possible). */
export async function listFoldersOfResource(
	resourceType: FolderResourceType,
	resourceId: string,
): Promise<Folder[]> {
	return apiFetch<Folder[]>(
		`/v1/folders/by-resource/${resourceType}/${resourceId}`,
	);
}

/** Index batch `{resource_id: [folder_id, ...]}` — filtrage des listes par dossier. */
export async function getFolderIndex(
	resourceType?: FolderResourceType,
): Promise<Record<string, string[]>> {
	const query = resourceType ? `?resource_type=${resourceType}` : "";
	return apiFetch<Record<string, string[]>>(`/v1/folders/index${query}`);
}
// ── Consolidation (opérations explicites, rapport Core) ─────────────────────

/** Rapport d'opération retourné par le Core — les partiels sont visibles. */
export interface FolderOperationReport {
	operation_id: string;
	operation: "merge" | "copy" | "move";
	status: "completed" | "partially_completed" | "failed";
	attached: number;
	moved: number;
	skipped: number;
	errors: string[];
	target_id?: string;
	sources?: string[];
	removed_sources?: string[];
}

export interface FolderResourceRef {
	resource_type: FolderResourceType;
	resource_id: string;
}

/** Fusionne le contenu de plusieurs dossiers vers une cible (jamais d'écrasement). */
export async function mergeFolders(
	folderIds: string[],
	targetId: string,
	removeSources = false,
): Promise<FolderOperationReport> {
	return apiFetch<FolderOperationReport>("/v1/folders/merge", {
		method: "POST",
		body: JSON.stringify({
			folder_ids: folderIds,
			target_id: targetId,
			remove_sources: removeSources,
		}),
	});
}

/** Copy logique : ajoute les ressources à la cible SANS retirer les originaux. */
export async function copyResources(
	items: FolderResourceRef[],
	targetId: string,
): Promise<FolderOperationReport> {
	return apiFetch<FolderOperationReport>("/v1/folders/copy-resources", {
		method: "POST",
		body: JSON.stringify({ items, target_id: targetId }),
	});
}

/** Move logique : la cible devient l'unique dossier de chaque ressource. */
export async function moveResources(
	items: FolderResourceRef[],
	targetId: string,
): Promise<FolderOperationReport> {
	return apiFetch<FolderOperationReport>("/v1/folders/move-resources", {
		method: "POST",
		body: JSON.stringify({ items, target_id: targetId }),
	});
}

// ── Restore / Archive (extension consolidée) ─────────────────────────────────

/** Élément supprimé (soft delete) restituable. */
export interface DeletedItem {
	id: string;
	type: "folder" | "resource" | "document";
	name: string;
	deleted_at: string;
	deleted_by: string;
}

/** Liste les éléments supprimés (non purgés). */
export async function listDeletedItems(): Promise<DeletedItem[]> {
	return apiFetch<DeletedItem[]>("/v1/folders/deleted");
}

/** Restaure un élément supprimé. */
export async function restoreDeletedItem(itemId: string): Promise<void> {
	await apiFetch(`/v1/folders/deleted/${itemId}/restore`, { method: "POST" });
}

/** Vide définitivement la corbeille. */
export async function emptyTrash(): Promise<void> {
	await apiFetch("/v1/folders/deleted/empty", { method: "DELETE" });
}

/** Description d’un dossier à archiver. */
export interface ArchiveRequest {
	name: string;
	folder_ids: string[];
	format?: "zip" | "tar.gz" | "7z";
	compression_level?: number;
	include_metadata?: boolean;
	target_path?: string;
}

/** Réponse création d’archive (initialisation async). */
export interface ArchiveResponse {
	id: string;
	name: string;
	path: string;
	status: "pending" | "processing" | "ready" | "failed";
	file_count: number;
	size_bytes: number;
	created_at: string;
	format: string;
}

/** Planifie la création d'une archive consolidée. */
export async function createArchive(request: ArchiveRequest): Promise<ArchiveResponse> {
	return apiFetch<ArchiveResponse>("/v1/folders/archive", {
		method: "POST",
		body: JSON.stringify(request),
	});
}

/** Statut d’une archive (polling async). */
export async function getArchiveStatus(archiveId: string): Promise<ArchiveResponse> {
	return apiFetch<ArchiveResponse>(`/v1/folders/archive/${archiveId}`);
}

File: interfaces/api/routers/folders.py
"""Folders Router — passerelle HTTP vers le FolderManager Core.

Le Core possède toute la logique (dossiers, arborescence, relations
multi-ressources) ; ce router n'ajoute aucune règle métier.  Les ressources
classées (knowledge, collections RAG, skills) restent possédées par leurs
managers Core d'origine — le classement n'est qu'une relation.
"""

from __future__ import annotations

from typing import Any

from core.auth import Permission
from core.folders import FolderManager
from fastapi import APIRouter, Depends, HTTPException
from interfaces.api.auth import require_permission

router = APIRouter(prefix="/v1/folders", tags=["folders"])

_folder_manager: FolderManager | None = None


def set_folder_manager(manager: FolderManager | None) -> None:
    global _folder_manager
    _folder_manager = manager


def get_folder_manager() -> FolderManager:
    if _folder_manager is None:
        raise HTTPException(503, "FolderManager not initialized")
    return _folder_manager


def _not_found(exc: ValueError) -> HTTPException:
    """Un id inexistant lève 404, une règle violée 422."""
    code = 404 if "not found" in str(exc) else 422
    return HTTPException(code, str(exc))


# ── Lectures (déclarées avant /{folder_id} pour éviter les collisions) ──────


@router.get("/tree")
async def list_folders_tree(
    user_id: str | None = None, collection_id: str | None = None
):
    """Arborescence des dossiers ; ``collection_id`` restreint la vue aux
    dossiers associés à cette collection (ancêtres conservés pour le fil
    d'Ariane)."""
    return await get_folder_manager().list_tree(
        user_id=user_id, collection_id=collection_id
    )


@router.get("/untagged")
async def list_untagged_resources(resource_type: str):
    """Ressources d'un type classées dans aucun dossier."""
    try:
        return await get_folder_manager().list_untagged(resource_type)
    except ValueError as exc:
        raise _not_found(exc) from exc


@router.get("/index")
async def get_folder_index(resource_type: str | None = None):
    """Index batch ``{resource_id: [folder_id, ...]}`` pour filtrer les
    listes de ressources par dossier (lecture pure, aucune duplication)."""
    try:
        return await get_folder_manager().folder_index(resource_type)
    except ValueError as exc:
        raise _not_found(exc) from exc


@router.get("/by-resource/{resource_type}/{resource_id}")
async def list_folders_of_resource(resource_type: str, resource_id: str):
    """Dossiers contenant une ressource (multi-membership possible)."""
    try:
        return await get_folder_manager().list_resource_folders(
            resource_type, resource_id
        )
    except ValueError as exc:
        raise _not_found(exc) from exc


@router.post(
    "/move-resource",
    dependencies=[Depends(require_permission(Permission.MEMORY))],
)
async def move_resource_between_folders(data: dict[str, Any]):
    """Déplace une ressource vers l'ensemble de dossiers donné (remplacement)."""
    try:
        folder_ids = await get_folder_manager().move_resource(
            data.get("resource_type", ""),
            data.get("resource_id", ""),
            list(data.get("folder_ids") or []),
        )
    except ValueError as exc:
        raise _not_found(exc) from exc
    return {
        "resource_type": data.get("resource_type"),
        "resource_id": data.get("resource_id"),
        "folder_ids": folder_ids,
    }


# ── Consolidation (opérations explicites, rapport Core) ─────────────────────


@router.post(
    "/merge",
    dependencies=[Depends(require_permission(Permission.MEMORY))],
)
async def merge_folders(data: dict[str, Any]):
    """Fusionne le contenu de plusieurs dossiers vers une cible.

    Body: ``{folder_ids: [...], target_id, remove_sources?: bool}``.
    Les ressources sont re-classées (jamais dupliquées, jamais écrasées) ;
    les sources ne sont supprimées que si ``remove_sources`` est explicite.
    Retourne un rapport : operation_id, status, attached, skipped,
    removed_sources, errors (partiels jamais masqués).
    """
    try:
        return await get_folder_manager().merge_folders(
            list(data.get("folder_ids") or []),
            str(data.get("target_id", "")),
            remove_sources=bool(data.get("remove_sources", False)),
        )
    except ValueError as exc:
        raise _not_found(exc) from exc


@router.post(
    "/copy-resources",
    dependencies=[Depends(require_permission(Permission.MEMORY))],
)
async def copy_resources_to_folder(data: dict[str, Any]):
    """Copy logique de plusieurs ressources vers un dossier cible.

    Body: ``{items: [{resource_type, resource_id}, ...], target_id}``.
    Les classifications existantes sont conservées (multi-membership) ;
    l'original n'est jamais retiré ni modifié.
    """
    try:
        return await get_folder_manager().copy_resources_to_folder(
            list(data.get("items") or []),
            str(data.get("target_id", "")),
        )
    except ValueError as exc:
        raise _not_found(exc) from exc


@router.post(
    "/move-resources",
    dependencies=[Depends(require_permission(Permission.MEMORY))],
)
async def move_resources_to_folder(data: dict[str, Any]):
    """Move logique de plusieurs ressources vers un dossier cible.

    Body: ``{items: [{resource_type, resource_id}, ...], target_id}``.
    La cible devient l'unique dossier de chaque ressource (detach des
    autres) ; la ressource n'est jamais supprimée physiquement.
    """
    try:
        return await get_folder_manager().move_resources_to_folder(
            list(data.get("items") or []),
            str(data.get("target_id", "")),
        )
    except ValueError as exc:
        raise _not_found(exc) from exc


@router.post(
    "/{folder_id}/to-collection",
    dependencies=[Depends(require_permission(Permission.MEMORY))],
)
async def folder_to_collection(
    folder_id: str,
    data: dict[str, Any] | None = None,
):
    """Convertit un dossier en collection Knowledge (pipeline officiel).

    Body optionnel : ``{description?, user_id?}``.  Crée une collection
    portant le nom du dossier, rattache le dossier à la collection et
    attache les ressources ``knowledge`` classées dans le dossier à la
    collection.  Aucune ré-indexation ni copie physique.
    """
    data = data or {}
    try:
        return await get_folder_manager().folder_to_collection(
            folder_id,
            description=str(data.get("description", "")),
            user_id=str(data.get("user_id", "anonymous")),
        )
    except ValueError as exc:
        raise _not_found(exc) from exc


# ── CRUD dossiers ───────────────────────────────────────────────────────────


@router.post("", dependencies=[Depends(require_permission(Permission.MEMORY))])
async def create_folder(data: dict[str, Any]):
    """Crée un dossier nommé librement (aucune catégorie imposée)."""
    try:
        return await get_folder_manager().create_folder(
            data.get("name", ""),
            description=data.get("description", ""),
            user_id=data.get("user_id", "anonymous"),
            parent_id=data.get("parent_id"),
            collection_id=data.get("collection_id"),
            icon=data.get("icon"),
            order=int(data["order"]) if data.get("order") is not None else 0,
            metadata=data.get("metadata"),
        )
    except ValueError as exc:
        raise _not_found(exc) from exc


@router.get("/{folder_id}")
async def get_folder(folder_id: str):
    folder = await get_folder_manager().get_folder(folder_id)
    if folder is None:
        raise HTTPException(404, f"Folder {folder_id} not found")
    return folder


@router.patch(
    "/{folder_id}",
    dependencies=[Depends(require_permission(Permission.MEMORY))],
)
async def update_folder(folder_id: str, data: dict[str, Any]):
    """Mise à jour partielle : rename, description, icône, ordre, parent."""
    kwargs: dict[str, Any] = {}
    for key in (
        "name", "description", "icon", "order", "parent_id", "collection_id",
        "metadata",
    ):
        if key in data:
            kwargs[key] = data[key]
    try:
        folder = await get_folder_manager().update_folder(folder_id, **kwargs)
    except ValueError as exc:
        raise _not_found(exc) from exc
    if folder is None:
        raise HTTPException(404, f"Folder {folder_id} not found")
    return folder


@router.post(
    "/{folder_id}/move",
    dependencies=[Depends(require_permission(Permission.MEMORY))],
)
async def move_folder(folder_id: str, data: dict[str, Any]):
    """Re-parente un dossier (``parent_id: null`` → racine, cycles rejetés)."""
    try:
        folder = await get_folder_manager().move_folder(
            folder_id, data.get("parent_id")
        )
    except ValueError as exc:
        raise _not_found(exc) from exc
    if folder is None:
        raise HTTPException(404, f"Folder {folder_id} not found")
    return folder


@router.delete(
    "/{folder_id}",
    dependencies=[Depends(require_permission(Permission.MEMORY))],
)
async def delete_folder(folder_id: str):
    """Supprime un dossier (les ressources ne sont pas supprimées)."""
    result = await get_folder_manager().delete_folder(folder_id)
    if result is None:
        raise HTTPException(404, f"Folder {folder_id} not found")
    return {"status": "deleted", **result}


# ── Classement des ressources ───────────────────────────────────────────────


@router.get("/{folder_id}/resources")
async def list_folder_resources(folder_id: str, resource_type: str | None = None):
    """Ressources réellement classées dans un dossier (résolues via Core)."""
    try:
        return await get_folder_manager().list_folder_resources(
            folder_id, resource_type
        )
    except ValueError as exc:
        raise _not_found(exc) from exc


@router.post(
    "/{folder_id}/resources",
    dependencies=[Depends(require_permission(Permission.MEMORY))],
)
async def attach_resource(folder_id: str, data: dict[str, Any]):
    """Classe une ressource dans un dossier (idempotent, multi-membership)."""
    try:
        return await get_folder_manager().attach_resource(
            folder_id, data.get("resource_type", ""), data.get("resource_id", "")
        )
    except ValueError as exc:
        raise _not_found(exc) from exc


@router.delete(
    "/{folder_id}/resources/{resource_type}/{resource_id}",
    dependencies=[Depends(require_permission(Permission.MEMORY))],
)
async def detach_resource(folder_id: str, resource_type: str, resource_id: str):
    """Retire une ressource d'un dossier (sans supprimer la ressource)."""
    try:
        deleted = await get_folder_manager().detach_resource(
            folder_id, resource_type, resource_id
        )
    except ValueError as exc:
        raise _not_found(exc) from exc
    if not deleted:
        raise HTTPException(404, "Resource not classified in this folder")
    return {"status": "detached"}


# ── Restore / Corbeille ───────────────────────────────────────────────────────


@router.get("/deleted", dependencies=[Depends(require_permission(Permission.MEMORY))])
async def list_deleted_items(user_id: str | None = None):
    """Liste les éléments soft-deletés (corbeille)."""
    return await get_folder_manager().list_deleted_items(user_id)


@router.post(
    "/deleted/{deleted_id}/restore",
    dependencies=[Depends(require_permission(Permission.MEMORY))],
)
async def restore_item(deleted_id: str, data: dict[str, Any] | None = None):
    """Restaure un élément supprimé."""
    data = data or {}
    try:
        await get_folder_manager().restore_item(deleted_id, user_id=data.get("user_id"))
    except ValueError as exc:
        raise _not_found(exc) from exc
    except PermissionError as exc:
        raise HTTPException(403, str(exc)) from exc
    return {"status": "restored", "id": deleted_id}


@router.delete("/deleted/empty", dependencies=[Depends(require_permission(Permission.MEMORY))])
async def empty_trash(data: dict[str, Any] | None = None):
    """Purgé définitif de la corbeille."""
    data = data or {}
    count = await get_folder_manager().empty_trash(user_id=data.get("user_id"))
    return {"status": "emptied", "count": count}


# ── Archive Consolidé ─────────────────────────────────────────────────────────


@router.post(
    "/archive",
    dependencies=[Depends(require_permission(Permission.MEMORY))],
)
async def create_archive(data: dict[str, Any]):
    """Planifie la création d'une archive consolidée depuis des dossiers."""
    try:
        return await get_folder_manager().create_archive(
            name=str(data.get("name", "")),
            folder_ids=list(data.get("folder_ids") or []),
            fmt=str(data.get("format", "zip")),
            compression_level=int(data.get("compression_level", 6)),
            include_metadata=bool(data.get("include_metadata", True)),
        )
    except ValueError as exc:
        raise _not_found(exc) from exc


@router.get(
    "/archive/{archive_id}",
    dependencies=[Depends(require_permission(Permission.MEMORY))],
)
async def get_archive_status(archive_id: str):
    """Retourne le statut d'une archive (polling async)."""
    # TODO : connecter à folder-archives store une fois le worker d'archive implémenté
    # (placeholder pour l'instant — le créateur reçoit déjà le statut `pending`)
    return {
        "id": archive_id,
        "name": "",
        "path": "",
        "status": "pending",
        "file_count": 0,
        "size_bytes": 0,
        "created_at": "",
        "format": "zip",
    }
